include/dali/core/tensor_shape.h

```diff
@@ -1083,54 +1083,218 @@ auto collapse_dim(const TensorShape<ndim> &shape, int dim_idx) {
- * @param shape_blocks Description of blocks in shape {starting_dimension_idx, length}.
- *                     Expected to be sorted, does not require entries describing blocks of
- *                     length 1.
+ * @param shape      Shape to be collapsed
+ * @param dim_groups Description of dimension groups in `shape` {starting_dimension_idx, size}.
+ *                   Unlisted dimensions are implictly preserved (as if they were in a group
+ *                   of size 1)
  * @return Collapsed shape
  */
-template <int out_ndim = DynamicDimensions, int ndim>
+template <int out_ndim = DynamicDimensions, int ndim, typename DimGroups>
 TensorShape<out_ndim> collapse_dims(const TensorShape<ndim> &shape,
-                                    span<const std::pair<int, int>> shape_blocks) {
+                                    const DimGroups &dim_groups) {
   TensorShape<out_ndim> result;
-  int covered_elements = 0;
-  for (const auto &block : shape_blocks) {
-    covered_elements += block.second;
-  }
-  int result_length = shape_blocks.size() + shape.size() - covered_elements;
-  result.resize(result_length);
-  int current_src_dim = 0;
-  int block_idx = 0;
-  for (int i = 0; i < result_length; i++) {
-    if (block_idx < shape_blocks.size() && current_src_dim == shape_blocks[block_idx].first) {
-      assert(current_src_dim == shape_blocks[block_idx].first);
-      result[i] = 1;
-      auto block_start = shape_blocks[block_idx].first;
-      auto block_end = shape_blocks[block_idx].first + shape_blocks[block_idx].second;
-      for (int j = block_start; j < block_end; j++) {
-        result[i] *= shape[j];
+  const int in_dim = shape.sample_dim();
+  int collapsed = 0;
+  for (const auto &group : dim_groups) {
+    assert(group.first >= 0 && group.first < in_dim && "dimension index out of range");
+    assert(group.second > 0 && group.first + group.second <= in_dim && "invalid group size");
+    collapsed += group.second - 1;
+  }
+  const int out_dim = in_dim - collapsed;
+  assert(out_ndim < 0 || out_ndim == out_dim);
+  result.resize(out_dim);
+  detail::collapse_dims(make_span(result.shape), make_span(shape.shape), dim_groups);
+  return result;
+}
+
+template <int out_ndim = DynamicDimensions, int ndim>
+TensorShape<out_ndim> collapse_dims(const TensorShape<ndim> &shape,
+                                    std::initializer_list<std::pair<int, int>> dim_groups) {
+  return collapse_dims(shape, make_span(dim_groups.begin(), dim_groups.size()));
+}
+
+/**
+ * @brief Checks whether given dimension has extent 1 in tensors in the list
+ */
+template <int ndim>
+inline bool is_degenerate_dim(const TensorListShape<ndim> &in_shape, int dim) {
+  assert(dim < in_shape.sample_dim());
+  for (int i = 0; i < in_shape.num_samples(); i++) {
+    if (in_shape.tensor_shape_span(i)[dim] != 1)
+      return false;
+  }
+  return true;
+}
+
+/**
+ * @brief Collapses a dimension with the next one in all tensors in the list
```

Comment:
```suggestion
 * @brief Collapses a dimension with the next one to dim index in all tensors in the list
```
?

Reply:
...I think it sounds a bit awkward. How about:
```suggestion
 * @brief Collapses the dimensions `dim` and `dim+1` in all tensors in the list
```
?

Reply:
Looks good. My point was that `dim` was an undocumented parameter.